### PR TITLE
CAT-2347 Rewrite BroadcastSequence(Wait)Map as a member of Sprite

### PR DESCRIPTION
**Pull request by rschluesselbauer**
Make BroadcastSequenceMap and BroadcastWaitSequenceMap a member of each sprite, adjust Broadcast unit tests, rewrite Broadcast regression tests in espresso

- BroadcastSequenceMap is changed to be a member of Sprite instead
  of one static map (this way we know which Sprite should react on
  which Broadcast)
- BroadcastWaitSequenceMap is changed to be a member of Sprite instead
  of one static map (for consistency with BroadcastSequenceMap), it now maps from Sprite to BroadcastSequenceMap (this is done so we know which sprites we want to do which actions)
- Updated calls to new Broadcast(Wait)SequenceMap in BroadcastHandler and StageDialog
- Removed for-loop from BroadcastHandler method handleActionFromBroadcastWait(..) that did not loop
- Regression Tests for Broadcasts are rewritten in espresso
- Added ProjectSetupUtils for putting together test projects more easily
- Added methods to UiTestUtils to efficiently wait for a UserVariable to have a specific value

### CODE REVIEW

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/UiTestUtils.java [hunk 1]
```diff
@@ -203,4 +206,28 @@ public static boolean comparePixelRgbaArrays(byte[] firstArray, byte[] secondArr
 		}
 		return true;
 	}
+
+	public static void waitForVariableValueEquals(UserVariable userVariable, double expectedValue, int timeout) {
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/UiTestUtils.java]: istead of the hard to read and comprehend version with the return; and assert in bottom, wouldnt it be nicer to have this function return a boolean, and rename it to something like "userVariableEqualsWithinTimeout" to then do something like:
assertTrue(userVariableEqualsValueWithinTimeout(uv, ev, t))

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/UiTestUtils.java [hunk 2]
```diff
@@ -203,4 +206,28 @@ public static boolean comparePixelRgbaArrays(byte[] firstArray, byte[] secondArr
...
+			if (Math.abs((double) userVariable.getValue() - expectedValue) < epsilon) {
+				return;
+			}
+			onView(isRoot()).perform(CustomActions.wait(step));
+		}
+		Assert.assertEquals(expectedValue, (double) userVariable.getValue(), epsilon);
+	}
+
+	public static void waitForVariableValueGreaterThan(UserVariable userVariable, double expectedValue, int timeout) {
+		double epsilon = 0.001;
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/UiTestUtils.java]: make epsilon a constant member of the utils class

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/UiTestUtils.java [hunk 3]
```diff
@@ -203,4 +206,28 @@ public static boolean comparePixelRgbaArrays(byte[] firstArray, byte[] secondArr
...
+		int step = 10;
+		for (; timeout > 0; timeout -= step) {
+			if ((double) userVariable.getValue() > (expectedValue + epsilon)) {
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/UiTestUtils.java]: shouldnt that be (expectedValue - epsilon)?

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/ProjectSetupUtils.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+		return broadcastScript;
+	}
+
+	public static Script addSpriteWithScript(Project project, String spriteName, Script script) {
+		Sprite sprite = new Sprite(spriteName);
+		project.getDefaultScene().addSprite(sprite);
+		return addScriptToSprite(script, sprite);
+	}
+
+	public static Script addScriptToSprite(Script script, Sprite sprite) {
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/ProjectSetupUtils.java]: why tho?
i dont see the point of this function, caller has to know script and sprite and can just call sprite.addScript himself.

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/ProjectSetupUtils.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+		SetVariableBrick setVariableBrickAfterBroadcast = new SetVariableBrick(new Formula(value), userVariable);
+		broadcastScript.addBrick(setVariableBrickAfterBroadcast);
+		sprite.addScript(broadcastScript);
+		return broadcastScript;
+	}
+
+	public static Script addSpriteWithScript(Project project, String spriteName, Script script) {
+		Sprite sprite = new Sprite(spriteName);
+		project.getDefaultScene().addSprite(sprite);
+		return addScriptToSprite(script, sprite);
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/ProjectSetupUtils.java]: see comment at function definition below.
this does not make it easier to see whats going on, especially with statements like this.

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/ProjectSetupUtils.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
+/*
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/ProjectSetupUtils.java]: since all of the functions in this class are very specifically designed for the two new tests, move the class over to /stage/ and rename to StageTestUtils. (eg. theres also a BrickTestUtils that also contains project setup stuff but again has different requirements than those for the stage)

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/ProjectSetupUtils.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+	public static Script addBroadcastScriptSettingVariable(Sprite sprite, String message, UserVariable userVariable,
+			double value) {
+		Script broadcastScript = new BroadcastScript(message);
+		SetVariableBrick setVariableBrickAfterBroadcast = new SetVariableBrick(new Formula(value), userVariable);
+		broadcastScript.addBrick(setVariableBrickAfterBroadcast);
+		sprite.addScript(broadcastScript);
+		return broadcastScript;
+	}
+
+	public static Script addSpriteWithScript(Project project, String spriteName, Script script) {
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/ProjectSetupUtils.java]: rework your function names, that addSpriteWithScript returns a Script is not at all clear from its name.

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/MultipleBroadcastsTest.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+		baseActivityTestRule.launchActivity(null);
+		Thread.sleep(1500);
+
+		for (int i = 0; i < variables.size(); i++) {
+			UiTestUtils.waitForVariableValueEquals(variables.get(i), values[i], 2000);
+		}
+	}
+
+	private List<UserVariable> createProject(int[] values) {
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/MultipleBroadcastsTest.java]: please rename this function. for the reader its not at all clear that "createProject" returns a List  of UserVariables.

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/MultipleBroadcastsTest.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+		Project project = UiTestUtils.createEmptyProject("test");
+		DataContainer dataContainer = project.getDefaultScene().getDataContainer();
+
+		ArrayList<UserVariable> userVariables = new ArrayList<>();
+		for (int i = 0; i < values.length; i++) {
+			Sprite sprite = new SingleSprite("sprite" + i);
+			project.getDefaultScene().addSprite(sprite);
+			UserVariable userVariable = dataContainer.addProjectUserVariable("var" + i);
+
+			Script startScript1 = ProjectSetupUtils.addScriptToSprite(new StartScript(), sprite);
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/MultipleBroadcastsTest.java]: please rename startScript**1**

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/MultipleBroadcastsTest.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+		baseActivityTestRule.launchActivity(null);
+		Thread.sleep(1500);
+
+		for (int i = 0; i < variables.size(); i++) {
+			UiTestUtils.waitForVariableValueEquals(variables.get(i), values[i], 2000);
+		}
+	}
+
+	private List<UserVariable> createProject(int[] values) {
+		Project project = UiTestUtils.createEmptyProject("test");
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/MultipleBroadcastsTest.java]: it would probably be nicer to have your own createEmptyProject function here, where you already get the data container so you dont have to dig for it.

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/MultipleBroadcastsTest.java [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+	}
+
+	@Test
+	public void testSendMultipleBroadcasts() throws InterruptedException {
+
+		int[] values = new int[] {30, 60, -30, -60};
+		List<UserVariable> variables = createProject(values);
+
+		baseActivityTestRule.launchActivity(null);
+		Thread.sleep(1500);
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/MultipleBroadcastsTest.java]: is this necessary?
if this is still required despite the waitForVariableValueEquals function below, maybe the timeout for it is too short, or you need to still need to add a idlingResource to the stage if this is due to boot up time of it. (i doubt that, it works on the simpleStageTest without anything.

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/ProjectSetupUtils.java [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+import org.catrobat.catroid.formulaeditor.Formula;
+import org.catrobat.catroid.formulaeditor.UserVariable;
+
+public final class ProjectSetupUtils {
+	// Suppress default constructor for noninstantiability
+	private ProjectSetupUtils() {
+		throw new AssertionError();
+	}
+
+	public static Script addBroadcastScriptSettingVariable(Sprite sprite, String message, UserVariable userVariable,
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/ProjectSetupUtils.java]: rename this function

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/UiTestUtils.java [hunk 13]
```diff
@@ -203,4 +206,28 @@ public static boolean comparePixelRgbaArrays(byte[] firstArray, byte[] secondArr
 		}
 		return true;
 	}
+
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/UiTestUtils.java]: please move both functions over to a .../stage/StageTestUtils.java class.

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/BroadcastReceiverRegressionTest.java [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+	private Project project;
+
+	@Rule
+	public BaseActivityInstrumentationRule<StageActivity> baseActivityTestRule = new
+			BaseActivityInstrumentationRule<>(StageActivity.class, true, false);
+
+	@Before
+	public void setUp() throws Exception {
+		BroadcastHandler.clearActionMaps();
+		project = UiTestUtils.createEmptyProject("test");
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/BroadcastReceiverRegressionTest.java]: make your own function, so you dont have to dig for the userVariable, and since you add a sprite and a start script in all of the below tests include it there aswell.

File: catroid/src/main/java/org/catrobat/catroid/common/BroadcastSequenceMap.java [hunk 15]
```diff
@@ -28,43 +28,43 @@
...
+		return broadcastSequenceMap.containsKey(sceneName) && broadcastSequenceMap.get(sceneName).containsKey(key);
 	}
 
-	public static boolean containsKey(String key, String sceneName) {
-		if (!BroadcastSequenceMap.broadcastSequenceMap.containsKey(sceneName)) {
-			return false;
+	public ArrayList<SequenceAction> get(String key, String sceneName) {
+		Map<String, ArrayList<SequenceAction>> map = broadcastSequenceMap.get(sceneName);
+		if (map == null) {
+			map = new HashMap<>();
```
Comment (84n4n4) [inline catroid/src/main/java/org/catrobat/catroid/common/BroadcastSequenceMap.java]: is this change required?
before BroadcastSequenceMap behave similar to a normal java Map, returning null if key is not found... now creation of new items in map is handeled in "get".

File: catroid/src/main/java/org/catrobat/catroid/common/BroadcastSequenceMap.java [hunk 16]
```diff
@@ -28,43 +28,43 @@
...
-		return BroadcastSequenceMap.broadcastSequenceMap.get(sceneName).containsKey(key);
+		return map.get(key);
 	}
 
-	public static ArrayList<SequenceAction> get(String key, String sceneName) {
-		if (!BroadcastSequenceMap.broadcastSequenceMap.containsKey(sceneName)) {
-			return null;
-		}
-		return BroadcastSequenceMap.broadcastSequenceMap.get(sceneName).get(key);
+	public boolean add(String sceneName, String key, SequenceAction value) {
```
Comment (84n4n4) [inline catroid/src/main/java/org/catrobat/catroid/common/BroadcastSequenceMap.java]: why rename to add? up until now BroadcastSequenceMap although not extending Map kind of behaved like what one would expect from something called *Map.

File: catroid/src/main/java/org/catrobat/catroid/common/BroadcastSequenceMap.java [hunk 17]
```diff
@@ -28,43 +28,43 @@
...
-	public static void clear() {
-		BroadcastSequenceMap.broadcastSequenceMap.clear();
+	public void clear() {
+		broadcastSequenceMap.clear();
 	}
 
-	public static void clear(String sceneName) {
-		if (!BroadcastSequenceMap.broadcastSequenceMap.containsKey(sceneName)) {
+	public void clear(String sceneName) {
+		if (!broadcastSequenceMap.containsKey(sceneName)) {
```
Comment (84n4n4) [inline catroid/src/main/java/org/catrobat/catroid/common/BroadcastSequenceMap.java]: please change to:
if (broadcastSequenceMap.containsKey(sceneName)) {
    broadcastSequenceMap.get(sceneName).clear();
}
its a lot easier to read...

File: catroid/src/main/java/org/catrobat/catroid/common/BroadcastWaitSequenceMap.java [hunk 18]
```diff
@@ -25,68 +25,66 @@
...
-public final class BroadcastWaitSequenceMap {
-	private static Map<String, HashMap<String, ArrayList<SequenceAction>>> broadcastWaitSequenceMap = new HashMap<>();
-	private static BroadcastEvent currentBroadcastEvent = null;
+public class BroadcastWaitSequenceMap {
+	private BroadcastEvent currentBroadcastEvent = null;
+	private Map<Sprite, BroadcastSequenceMap> waitSequenceMap;
 
-	private BroadcastWaitSequenceMap() {
-		throw new AssertionError();
+	public BroadcastWaitSequenceMap() {
```
Comment (84n4n4) [inline catroid/src/main/java/org/catrobat/catroid/common/BroadcastWaitSequenceMap.java]: why initialization different than in BroadcastSequenceMap?

File: catroid/src/main/java/org/catrobat/catroid/common/BroadcastWaitSequenceMap.java [hunk 19]
```diff
@@ -25,68 +25,66 @@
...
+		waitSequenceMap = new HashMap<>();
 	}
 
-	public static boolean containsKey(String key, String sceneName) {
-		if (!broadcastWaitSequenceMap.containsKey(sceneName)) {
-			return false;
-		}
-		return BroadcastWaitSequenceMap.broadcastWaitSequenceMap.get(sceneName).containsKey(key);
+	public void add(String sceneName, String key, Sprite sprite, ArrayList<SequenceAction> values) {
```
Comment (84n4n4) [inline catroid/src/main/java/org/catrobat/catroid/common/BroadcastWaitSequenceMap.java]: see comments in BroadcastSequenceMap

File: catroid/src/main/java/org/catrobat/catroid/content/BroadcastHandler.java [hunk 20]
```diff
@@ -51,48 +51,52 @@ private BroadcastHandler() {
...
 			Script scriptOfAction = actionScriptMap.get(action);
 
 			if (!handleAction(action, scriptOfAction)) {
 				addOrRestartAction(look, action);
 			}
 		}
 
-		if (BroadcastWaitSequenceMap.containsKey(broadcastMessage, sceneName)) {
-			for (SequenceAction action : BroadcastWaitSequenceMap.get(broadcastMessage, sceneName)) {
+		if (event.getSenderSprite().getBroadcastWaitSequenceMap().containsKey(broadcastMessage, sceneName, look.sprite)) {
```
Comment (84n4n4) [inline catroid/src/main/java/org/catrobat/catroid/content/BroadcastHandler.java]: this long dig: event.getSenderSprite().getBroadcastWaitSequenceMap(). show up a lot.
maybe make a local variable for it and rename the broadcastSequenceMap and this new one so that its clear whos who.

File: catroid/src/main/java/org/catrobat/catroid/content/Look.java [hunk 21]
```diff
@@ -106,7 +106,7 @@ public boolean touchDown(InputEvent event, float x, float y, int pointer, int bu
 		this.addListener(new BroadcastListener() {
 			@Override
 			public void handleBroadcastEvent(BroadcastEvent event, String broadcastMessage) {
-				doHandleBroadcastEvent(broadcastMessage);
+				doHandleBroadcastEvent(event, broadcastMessage);
```
Comment (84n4n4) [inline catroid/src/main/java/org/catrobat/catroid/content/Look.java]: event is handed through a few functions just to be used for getting the map (event.getSenderSprite().getBroadcastWaitSequenceMap())
maybe pass the map through?

File: catroid/src/main/java/org/catrobat/catroid/content/Sprite.java [hunk 22]
```diff
@@ -379,8 +376,7 @@ public Sprite cloneForCloneBrick() {
 	}
 
 	public Sprite cloneForScene() {
-		Sprite clone = clone();
-		return clone;
+		return clone();
```
Comment (84n4n4) [inline catroid/src/main/java/org/catrobat/catroid/content/Sprite.java]: find usages of this functions and change them to call clone() directly and then remove this function

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/UiTestUtils.java (hunk 3 above)
Comment (rschluesselbauer) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/UiTestUtils.java]: I don't think so.
With greater than we want the two values to be at least epsilon (plus a value >0) apart.
Saying value > expectedValue-epsilon would have the effect that value could be smaller than expectedValue.

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/ProjectSetupUtils.java (hunk 4 above)
Comment (rschluesselbauer) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/ProjectSetupUtils.java]: It saves a line when creating scripts for a sprite like this:
`Script startScript = ProjectSetupUtils.addScriptToSprite(new StartScript(), sprite);`

I'll still remove it, as you are right and the benefit is minimal.

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/MultipleBroadcastsTest.java (hunk 11 above)
Comment (rschluesselbauer) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/MultipleBroadcastsTest.java]: Sorry, that's something I overlooked while refactoring. My bad.

File: catroid/src/main/java/org/catrobat/catroid/common/BroadcastSequenceMap.java (hunk 15 above)
Comment (rschluesselbauer) [inline catroid/src/main/java/org/catrobat/catroid/common/BroadcastSequenceMap.java]: Concerning the map changes - I think I got a bit carried away there.
I'll change it back to what you would expect of a map.
Thanks for the elaborate feedback!

**Comment by rschluesselbauer** [comment]
Thanks for the elaborate feedback!
I have incoorporated almost all of it, except for the one with epsilon (shouldnt that be (expectedValue - epsilon)?) in the method userVariableGreaterThanWithinTimeout(..) in StageTestUtils.
I also fixed the PhysicsCollisionBaseTest, as it didn't add sprite2 to the project before (resulting in a failing build)
I also added a nullcheck for senderSprite, as it is null in case of a CollisionBroadcast.

**Comment by 84n4n4** [comment]
hi!

regarding the epsilon thing. that depends, if you approach it mathematically:
when comparing two floats, two floats can be considered equal if they are closer than epsilon together. so when a is growing, and b is your requested limit, above condition would be reached when a == b - epsilon
on the other hand, if you really want to make sure that a > b then stating a > b is sufficient. (because in this code snippet, a could be already bigger than b while still within epsilon...)

**Comment by rschluesselbauer** [comment]
Ok, so when a is growing and b is the limit
a > (b - epsilon) is the same thing as a **greaterOrEqual** to b
a > (b + epsilon) is the same thing as a **greaterAndNotEqual** to b

I am going with the second one, because I don't want the condition to be true when a and b are equal (within the epsilon).

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/MultipleBroadcastsTest.java [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+	public void setUp() throws Exception {
+		BroadcastHandler.clearActionMaps();
+		createProject();
+	}
+
+	@Test
+	public void testSendMultipleBroadcasts() throws InterruptedException {
+
+		int[] values = new int[] {30, 60, -30, -60};
+		List<UserVariable> variables = setUpProjectAndReturnUserVariables(values);
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/MultipleBroadcastsTest.java]: now theres a createProject and a setUpProject.. function. since theres only one test here, please move everything that is actually setting up for this test into the setUp function. (and merge createProject and setUpProject... into one)

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/StageTestUtils.java [hunk 24]
```diff
@@ -0,0 +1,84 @@
...
+	}
+
+	public static Script addScriptToAnonymousSprite(Project project, String spriteName, Script script) {
+		Sprite sprite = new Sprite(spriteName);
+		project.getDefaultScene().addSprite(sprite);
+		sprite.addScript(script);
+		return script;
+	}
+
+	public static boolean userVariableEqualsWithinTimeout(UserVariable userVariable, double expectedValue, int timeout) {
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/StageTestUtils.java]: can you please rename timeout to timeoutMs or something like this to make it clear that its in ms.

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/StageTestUtils.java [hunk 25]
```diff
@@ -0,0 +1,84 @@
...
+		int step = 10;
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/StageTestUtils.java]: please rename step to checkIntervalMs (or something suitable that communicates that this is a time step and its in ms)

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/StageTestUtils.java [hunk 26]
```diff
@@ -0,0 +1,84 @@
...
+public final class StageTestUtils {
+
+	private static final double EPSILON = 0.001;
+
+	// Suppress default constructor for noninstantiability
+	private StageTestUtils() {
+		throw new AssertionError();
+	}
+
+	public static Script addBroadcastScriptSettingUserVariableToValue(Sprite sprite, String message, UserVariable userVariable,
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/StageTestUtils.java]: can you please refactor this.
if a function is called addAtoB i expect it to do just that without returning another object. so either refactor into a different function structure, or rename to make it clear what this does (if its getting hard to put a name on your functions because they do so much at once, splitting them up is maybe a good idea)

File: catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/StageTestUtils.java [hunk 27]
```diff
@@ -0,0 +1,84 @@
...
+			double value) {
+		Script broadcastScript = new BroadcastScript(message);
+		SetVariableBrick setVariableBrickAfterBroadcast = new SetVariableBrick(new Formula(value), userVariable);
+		broadcastScript.addBrick(setVariableBrickAfterBroadcast);
+		sprite.addScript(broadcastScript);
+		return broadcastScript;
+	}
+
+	public static Script addScriptToAnonymousSprite(Project project, String spriteName, Script script) {
```
Comment (84n4n4) [inline catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/stage/StageTestUtils.java]: same as last comment

**Comment by rschluesselbauer** [comment]
All suggested changes have been made. Anything else I should improve on?

**Comment by 84n4n4** [comment]
test this please

**Comment by 84n4n4** [comment]
@cheidenreich  can you please have a look.
for me this looks fine.

**Comment by cheidenreich** [comment]
thanks @84n4n4 for the review.

I had another look on the code and tested the functionality manually on my device, even with heavy copy/paste/delete nothing breaks, looks good

**Comment by cheidenreich** [comment]
test this please

**Comment by cheidenreich** [comment]
please run espresso tests

**Comment by cheidenreich** [comment]
test this please
